Dev/production doc (#824)
AMD-AGI/Primus: production_docs/01-getting-started/overview.md

# Primus overview

## Executive summary

> **Primus (Primus-LM)** is a YAML-driven training framework for large-scale foundation model work on AMD GPUs (ROCm). It targets **machine learning engineers**, **researchers**, and **platform/operations teams** who need reproducible, multi-backend training pipelines on AMD Instinct hardware.

Reviewer: It is not clear what is Primus-LM and how it is related to Primus. Either add some explanations, for remove Primus-LM.

> **Primus (Primus-LM)** is a YAML-driven training framework for large-scale foundation model work on AMD GPUs (ROCm). It targets **machine learning engineers**, **researchers**, and **platform/operations teams** who need reproducible, multi-backend training pipelines on AMD Instinct hardware.

Reviewer: ```suggestion
**Primus (Primus-LM)** is a YAML-driven training framework for large-scale foundation model work on AMD GPUs. It targets **machine learning engineers**, **researchers**, and **platform/operations teams** who need reproducible, multi-backend training pipelines on AMD Instinct hardware.
```

**Repository:** [https://github.com/AMD-AIG-AIMA/Primus](https://github.com/AMD-AIG-AIMA/Primus)

> **License:** The project README states Apache License 2.0; the repository root `LICENSE` file is MIT. Treat licensing as project-specific and confirm with your compliance process before redistribution.

Reviewer: This does not look to be the right place for discussing licensing. And it's also not clear when we have 2 licenses for this software, and what should users do with that.  Move this to e.g. end of the doc (?), and give clear explanations.

---

## What Primus provides

| Area | Description |
|------|-------------|
| **Multi-backend training** | One workflow surface over Megatron-LM, TorchTitan, JAX MaxText, Megatron Bridge, and HummingbirdXT. |
| **Unified CLI** | `primus-cli` with **direct** (bare metal), **container** (Docker/Podman), and **slurm** (cluster) execution modes. |
| **YAML-driven configuration** | Experiment, model, module, and platform presets composed from reusable fragments under `primus/configs/`. |

> | **Benchmark suite** | Built-in benchmarks (for example GEMM) for quick hardware and stack validation. |

Reviewer: ```suggestion
| **Benchmark suite** | Built-in benchmarks (for example, GEMM) for quick hardware and stack validation. |
```

| **Preflight diagnostics** | Cluster-oriented checks for host, GPU, and network health before long jobs. |
| **Performance projection** | Tools to estimate memory use and throughput without occupying a full cluster. |

> Workflows span **pretraining** and **post-training** (including SFT and LoRA). Some Megatron configuration files include RL-related parameters, but this doc set does not treat RL workflows as publication-ready until maintainers add examples and support guidance.

Reviewer: Shouldn't discuss there what is or is not publication ready and  when they will be published. Explain it better here why RL workflows are not published (without promising we will publish them).

---

> ## Primus ecosystem (conceptual)

Reviewer: ```suggestion
## Primus ecosystem
```

> Primus-LM sits between stability/platform services and low-level operators:

Reviewer: Why highlight Primus-LM here, while we are listing the other components as well?

```
                    +------------------+
                    |   Primus-SaFE    |
                    | (stability /     |
                    |  cluster mgmt)   |
                    +--------+---------+
                             |
                    +--------v---------+
                    |    Primus-LM     |
                    |  (this repo:     |
                    |   training)      |
                    +--------+---------+
                             |
                    +--------v---------+
                    |  Primus-Turbo    |
                    | (operators /     |
                    |  kernels)        |
                    +------------------+
```

> - **Primus-SaFE**: Stability and fault-tolerance oriented cluster management referenced by auxiliary tooling, but not documented here as an in-repo production integration.

Reviewer: ```suggestion
- **Primus-SaFE**: Stability and fault-tolerance oriented cluster management referenced by auxiliary tooling (not documented here).
```

Reviewer: Add more explanation if needed

> - **Primus-LM**: Training orchestration, backends, CLI, and configs (this repository).

Reviewer: ```suggestion
- **Primus-LM**: Training orchestration, backends, CLI, and configurations (maintained in the [AMD-AGI/Primus](https://github.com/AMD-AGI/Primus/) repository).
```

> - **Primus-Turbo**: High-performance operators (for example FlashAttention-style kernels, GEMM, collectives).

Reviewer: ```suggestion
- **Primus-Turbo**: High-performance operators (for example, FlashAttention-style kernels, GEMM, and collectives).
```

---

## Supported backends

| Backend | Typical use |
|---------|-------------|
| **Megatron-LM** | Broadest model coverage; default for many GPT-style and MoE recipes. |
| **TorchTitan** | PyTorch-native large-model training paths. |
| **JAX MaxText** | JAX/Flax training stacks aligned with MaxText. |
| **Megatron Bridge** | Post-training and bridge workflows on top of Megatron-related stacks. |

> | **HummingbirdXT** | Additional integrated training path where enabled by your deployment. |

Reviewer: ```suggestion
| **HummingbirdXT** | Additional integrated training path wheen enabled by your deployment. |
```

Backend choice is expressed in experiment YAML and resolved through Primus’s adapter layer (see [glossary](./glossary.md)).

---

## Supported hardware and stack

| Item | Requirement |
|------|----------------|

> | **GPUs** | AMD Instinct **MI300X**, **MI325X**, **MI355X** |

Reviewer: ```suggestion
| **GPUs** | AMD Instinct™ **MI300X**, **MI325X**, **MI355X** |
```

| **Platform** | **ROCm** (version **≥ 7.0** recommended) |
| **Container image (reference)** | `docker.io/rocm/primus:v26.3` |

Exact kernel and driver packages should match AMD’s documentation for your GPU SKU and ROCm release.

---

## Key dependencies

> Primus integrates with the following categories of software (non-exhaustive):

Reviewer: ```suggestion
Primus depends on the following categories of software (the list is non-exhaustive):
```

| Category | Examples |
|----------|----------|
| **Framework** | PyTorch (Megatron-LM, TorchTitan paths); JAX/Flax (MaxText path). |
| **AMD stack** | ROCm, RCCL, HipBLASLt (GEMM), GPU drivers. |
| **Execution** | Docker or Podman for container mode; Slurm for cluster mode. |
| **Observability & tooling** | **loguru** (logging), **Weights & Biases** (`wandb`) and other optional trackers (see `requirements.txt`). |

> Install specifics are covered in [installation](./installation.md).

Reviewer: ```suggestion
Install specifics are covered in [Installation and setup](./installation.md).
```

---

## Runtime model

At a high level, a run follows this pipeline:

1. **YAML experiment config** defines work group, modules, model preset, and overrides.
2. **`primus-cli`** selects execution mode (direct, container, or slurm) and forwards to the runner.

> 3. **Backend adapter** maps the resolved config to the target framework (Megatron-LM, TorchTitan, and so on).

Reviewer: ```suggestion
3. **Backend adapter** maps the resolved configuration to the target framework (Megatron-LM, TorchTitan, and so on).
```

4. **Distributed launch** typically uses **`torchrun`** (or the backend’s equivalent) to start workers across GPUs and nodes.

> For CLI shape and options, see [quickstart](./quickstart.md) and [CLI reference](../02-user-guide/cli-reference.md).

Reviewer: ```suggestion
For CLI shape and options, see [Quickstart](./quickstart.md) and [CLI reference](../02-user-guide/cli-reference.md).
```

---

## Repository layout (top level)

| Path | Role |
|------|------|
| `primus/` | Core library: configs, runtime, trainers, backend adapters, tools (including preflight). |
| `runner/` | CLI implementation, helpers, hooks, and launch glue. |
| `examples/` | End-to-end example YAML and recipes per backend and GPU SKU. |

> | `docs/` | Upstream project documentation (architecture, CLI guides, backends). |

Reviewer: Will we still have production_docs? If so, list here.

| `tests/` | Automated tests. |
| `tools/` | Auxiliary scripts and utilities. |
| `benchmark/` | Benchmark drivers and related assets. |
| `third_party/` | Vendored or submodule dependencies. |

---

## Next steps

- [Installation](./installation.md) — ROCm, Docker, pip, and Slurm setup.

> - [Quickstart](./quickstart.md) — Run a minimal training job in minutes.

Reviewer: Perhaps move this up, to be consistent with the order in the TOC?
